mkosi/__init__.py

```diff
@@ -3665,7 +3665,9 @@ def insert_partition(
 
             # Without this the entire blob will be read into memory which could exceed system memory
             with open(path, mode='wb') as path_fp:
-                os.sendfile(path_fp.fileno(), blob.fileno(), offset=0, count=blob_size)
+                # Chunking required because sendfile under Linux has a maximum copy size
+                for offset in range(0, blob_size, 0x7ffff000):
+                    os.sendfile(path_fp.fileno(), blob.fileno(), offset=offset, count=0x7ffff000)
```

Comment:
Looking at the cpython implementation this seems to have overflow issues on 32bit architectures, which is why they choose either the file size as the block size or `2**30`, whichever is smaller.

Reply:
Is it the offset or the count that overflows under 32-bit? Either way I'm open to suggestions if you send me through an example.

Reply:
It's the [count](https://github.com/python/cpython/blob/main/Lib/shutil.py#L135)

Reply:
Exactly. I think going with a default block size of `0x7ffff000` is fine, since sendfile won't do more than that anyway, it only precludes automatically picking up a larger block size if sendfile ever starts doing that. I guess that's why they go with a larger block size in the first place. It only means you need to to the while loop and look at sendfile's return.

I think we should do the cap to `2**30` for 32 bit systems, too.

Reply:
Why not make our lives easier knowing that we can't send more than `0x7ffff000` bytes regardless of 32 or 64 bit architecture and just cap at `2**30`? The difference is about 1MiB (i.e.: a 1MiB inefficiency).

Reply:
Isn't the difference a factor of two and about 1GiB?

Reply:
You're right, it is about 1GiB. Whoops on my part. Looking harder I'm not convinced we need to set the count to `2**30` on 32-bit systems. It seems it just needs to be below `2**31- 1` for 32-bit.

From https://bugs.python.org/issue38319 which was mentioned in `_fastcopy_sendfile()`:

*I understand that the error comes from the 4th sendfile() parameter: "count". The C code (of the Linux implementation) uses the "n" format for Py_ssize_t: [Python/getargs.c](https://github.com/python/cpython/blob/master/Python/getargs.c) calls PyLong_AsSsize_t().*

*On a 64-bit system, it's less likely to reach Py_ssize_t maximum value (max = `2**63-1`), but it's easy to reach on a 32-bit system (max = `2**31-1`).*

Reply:
@behrmann Did you get a change to review further?

Reply:
Sorry forgot to answer. I guess we should err on the same size and mirror cpython here, e.g.
```python
chunksize = 0x7ffff000 if is_64bit() else 2**30
offset = 0
while True:
    sent = os.sendfile(path_fp.fileno(), blob.fileno(), offset=offset, count=chunksize)
    if not sent:
        break
    offset += sent
```
for some not yet existing function `is_64bit`

Reply:
Done.